Disallow setting some built-in cfg via the command-line

Urgau opened the pull request:
This PR disallow users from setting some built-in cfg via set the command-line in order to prevent incoherent state, eg. `windows` cfg active but target is Linux based.

This implements MCP https://github.com/rust-lang/compiler-team/issues/610, with the caveat that we disallow cfgs no matter if they make sense or not, since I don't think it's useful to allow users to set a cfg that will be set anyway. It also complicates the implementation.

------

The `explicit_builtin_cfgs_in_flags` lint detects builtin cfgs set via the `--cfg` flag.

*(deny-by-default)*

### Example

```text
rustc --cfg unix
```

```rust,ignore (needs command line option)
fn main() {}
```

This will produce:

```text
error: unexpected `--cfg unix` flag
  |
  = note: config `unix` is only supposed to be controlled by `--target`
  = note: manually setting a built-in cfg can and does create incoherent behaviours
  = note: `#[deny(explicit_builtin_cfgs_in_flags)]` on by default
```

### Explanation

Setting builtin cfgs can and does produce incoherent behaviour, it's better to the use the appropriate `rustc` flag that controls the config. For example setting the `windows` cfg but on Linux based target.

-----

r? @petrochenkov
cc @jyn514

try-job: aarch64-apple
try-job: test-various
try-job: armhf-gnu
try-job: x86_64-msvc
try-job: x86_64-mingw
try-job: i686-msvc
try-job: i686-mingw
try-job: x86_64-gnu-llvm-17
try-job: dist-various-1

Urgau committed:
Disallow setting built-in cfgs via set the command-line

petrochenkov commented on compiler/rustc_session/src/config/cfg.rs:
```diff
@@ -379,3 +380,54 @@ impl CheckCfg {
+    //  - rustfmt
+    //  - overflow_checks
+    //  - debug_assertions
+    //  - ub_checks
```
All of these may produce incoherent behavior, from what I see.
E.g. `test` and `doctest` are used in standard library to disable some panic machinery, if they are passed incorrectly then things will break.
So I don't see any fundamental differences from the configs above and below.
(Except maybe rustfmt or clippy, because I'm not sure why they exist at all.)

Perhaps they should all produce lint warnings.
The warnings can be silenced with `-A`/`#[allow]` if the user really needs it.

Urgau replied:
Of course, I've changed the implementation to a deny-by-default lint and included all possible cfgs (I've left some links in the code to explain why some are excluded).

(I had to modify the `BufferedEarlyLint` struct to contain an optional span, I can submit this patch as a separate PR if desired)

petrochenkov commented on compiler/rustc_session/src/config/cfg.rs:
```diff
@@ -379,3 +380,54 @@ impl CheckCfg {
+    //  - debug_assertions
+    //  - ub_checks
+
+    for cfg in user_cfgs {
```
Can this be merged into `default_configuration` somehow to keep things synchronized?

Urgau replied:
It would probably be better to merge the logic in `fill_well_known` since we know there all the cfgs but `fill_well_known` doesn't do anything when `--check-cfg` isn't active which complicates the logic more; so to keep things simplify for now I would prefer to keep the logic somewhat separate (it's still the same file).

But out of concern I moved the function just above the `default_configuration` function, to hopefully make it more obvious that the function exist and I added the function to the guide (at the top of the file) for when adding a new cfg.

Urgau committed:
Make `Span` optional in `BufferedEarlyLint`

rust-log-analyzer commented:
The job **`x86_64-gnu-tools`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/9625801328/job/26551075795) [(plain)](https://github.com/rust-lang/rust/commit/4ebdd5cbf517c0dbce4b68f5159196b51582a6cf/checks/26551075795/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
   Compiling ui_test v0.23.0
   Compiling clippy v0.1.81 (/checkout/src/tools/clippy)
error: unexpected `--cfg test` flag
  |
  = note: config `test` is only supposed to be controlled by `--test`
  = note: see <https://github.com/rust-lang/rust/issues/xxxxx> for more information
  = note: `#[deny(unexpected_builtin_cfgs)]` on by default
error: could not compile `clippy` (test "compile-test") due to 1 previous error
warning: build failed, waiting for other jobs to finish...


```

</details>

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-17`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/9625952591/job/26551384280) [(plain)](https://github.com/rust-lang/rust/commit/37d63c6b8d8755d43d47f12bc0e7fd83746a5fa9/checks/26551384280/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
 - LegacyKeyValueFormat: "ENV key=value" should be used instead of legacy "ENV key value" format (line 46)
 - LegacyKeyValueFormat: "ENV key=value" should be used instead of legacy "ENV key value" format (line 49)
 - LegacyKeyValueFormat: "ENV key=value" should be used instead of legacy "ENV key value" format (line 61)
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-17]
---
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-17', '--enable-llvm-link-shared', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'change-id=99999999', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-17/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.thin-lto-import-instr-limit := 10
configure: change-id            := 99999999
---
failures:

---- [codegen] tests/codegen/aarch64-struct-align-128.rs#windows stdout ----

error in revision `windows`: compilation failed!
status: exit status: 1
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/codegen/aarch64-struct-align-128.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--cfg" "windows" "--check-cfg" "cfg(FALSE,linux,darwin,windows)" "-O" "-Cdebug-assertions=no" "--emit" "llvm-ir" "-C" "prefer-dynamic" "-o" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/aarch64-struct-align-128.windows/aarch64-struct-align-128.ll" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/aarch64-struct-align-128.windows/auxiliary" "--target" "aarch64-pc-windows-msvc"
--- stderr -------------------------------
--- stderr -------------------------------
error: unexpected `--cfg windows` flag
  = note: config `windows` is only supposed to be controlled by `--target`
  = note: config `windows` is only supposed to be controlled by `--target`
  = note: see <https://github.com/rust-lang/rust/issues/xxxxx> for more information
  = note: `#[deny(unexpected_builtin_cfgs)]` on by default

warning: `extern` block uses type `i128`, which is not FFI-safe
    |
    |
100 |     fn test_i128(a: I128, b: TransparentI128, c: WrappedI128);
    |                     ^^^^ not FFI-safe
    = note: 128-bit integers don't currently have a known stable ABI
    = note: `#[warn(improper_ctypes)]` on by default


warning: `extern` block uses type `i128`, which is not FFI-safe
    |
    |
100 |     fn test_i128(a: I128, b: TransparentI128, c: WrappedI128);
    |                              ^^^^^^^^^^^^^^^ not FFI-safe
    = note: 128-bit integers don't currently have a known stable ABI


warning: `extern` block uses type `i128`, which is not FFI-safe
    |
    |
100 |     fn test_i128(a: I128, b: TransparentI128, c: WrappedI128);
    |                                                  ^^^^^^^^^^^ not FFI-safe
    = note: 128-bit integers don't currently have a known stable ABI


warning: `extern` block uses type `i128`, which is not FFI-safe
    |
    |
127 |     fn test_packed(a: Packed, b: TransparentPacked, c: WrappedPacked);
    |                       ^^^^^^ not FFI-safe
    = note: 128-bit integers don't currently have a known stable ABI


warning: `extern` block uses type `i128`, which is not FFI-safe
    |
    |
127 |     fn test_packed(a: Packed, b: TransparentPacked, c: WrappedPacked);
    |                                  ^^^^^^^^^^^^^^^^^ not FFI-safe
    = note: 128-bit integers don't currently have a known stable ABI


warning: `extern` block uses type `i128`, which is not FFI-safe
    |
    |
127 |     fn test_packed(a: Packed, b: TransparentPacked, c: WrappedPacked);
    |                                                        ^^^^^^^^^^^^^ not FFI-safe
    = note: 128-bit integers don't currently have a known stable ABI

error: aborting due to 1 previous error; 6 warnings emitted
------------------------------------------
------------------------------------------


---- [codegen] tests/codegen/default-requires-uwtable.rs#WINDOWS stdout ----

error in revision `WINDOWS`: compilation failed!
status: exit status: 1
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/codegen/default-requires-uwtable.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--cfg" "windows" "--check-cfg" "cfg(FALSE,windows,android)" "-O" "-Cdebug-assertions=no" "--emit" "llvm-ir" "-C" "prefer-dynamic" "-o" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/default-requires-uwtable.WINDOWS/default-requires-uwtable.ll" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/default-requires-uwtable.WINDOWS/auxiliary" "-C" "panic=abort" "-Copt-level=0" "--target=x86_64-pc-windows-msvc"
--- stderr -------------------------------
--- stderr -------------------------------
error: unexpected `--cfg windows` flag
  = note: config `windows` is only supposed to be controlled by `--target`
  = note: config `windows` is only supposed to be controlled by `--target`
  = note: see <https://github.com/rust-lang/rust/issues/xxxxx> for more information
  = note: `#[deny(unexpected_builtin_cfgs)]` on by default
error: aborting due to 1 previous error
------------------------------------------



---- [codegen] tests/codegen/repr/transparent-sysv64.rs#windows stdout ----

error in revision `windows`: compilation failed!
status: exit status: 1
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/codegen/repr/transparent-sysv64.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--cfg" "windows" "--check-cfg" "cfg(FALSE,linux,apple,windows)" "-O" "-Cdebug-assertions=no" "--emit" "llvm-ir" "-C" "prefer-dynamic" "-o" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/repr/transparent-sysv64.windows/transparent-sysv64.ll" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/repr/transparent-sysv64.windows/auxiliary" "-O" "-C" "no-prepopulate-passes" "--target" "x86_64-pc-windows-msvc"
--- stderr -------------------------------
--- stderr -------------------------------
error: unexpected `--cfg windows` flag
  = note: config `windows` is only supposed to be controlled by `--target`
  = note: config `windows` is only supposed to be controlled by `--target`
  = note: see <https://github.com/rust-lang/rust/issues/xxxxx> for more information
  = note: `#[deny(unexpected_builtin_cfgs)]` on by default
error: aborting due to 1 previous error
------------------------------------------


```

</details>

Urgau commented:
I've changed the error from a hard-error to a deny-by-default lint as asked and included the maximum of cfgs and added some links for the one that we can't lint on them.

@rustbot ready

petrochenkov commented on tests/codegen/default-requires-uwtable.rs:
```diff
@@ -1,9 +1,9 @@
-//@ revisions: WINDOWS ANDROID
+//@ revisions: WINDOWS_ ANDROID_
```
Why do these ones need to be renamed? They are not builtin.

Urgau replied:
bc compiletest normalize them to lowercase cfg, cf. https://github.com/rust-lang/rust/blob/2c243d957008f5909f7a4af19e486ea8a3814be7/src/tools/compiletest/src/runtest.rs#L682-L684

petrochenkov commented on tests/debuginfo/thread-names.rs:
```diff
@@ -1,8 +1,8 @@
 //@ compile-flags:-g
-//@ revisions: macos windows
+//@ revisions: macos windows_
```
```suggestion
//@ revisions: macos win
```
(For consistency with other tests.)

petrochenkov replied:
Also, maybe `#![allow]` would work better for cases like this?

Urgau replied:
> Also, maybe `#![allow]` would work better for cases like this?

I have a slight preference to not disturb test with unnecessary things, like allow of lints, strait lines, ... and be as straight as to the point.

petrochenkov commented on compiler/rustc_session/src/config/cfg.rs:
```diff
@@ -83,6 +87,76 @@ impl<'a, T: Eq + Hash + Copy + 'a> Extend<&'a T> for ExpectedValues<T> {
+        };
+        sess.psess.opt_span_buffer_lint(
+            UNEXPECTED_BUILTIN_CFGS,
+            None,
```
What happens if you use `DUMMY_SP` here instead of `None`?

Urgau replied:
I though putting `DUMMY_SP` as the primary span would be an un-allowed case (resulting in an ICE) but apparently it's only the case when putting dummy spans for suggestions.

Currently putting a dummy span forces our default error emitter to attach the diagnostic to the crate outermost file, which implies having `ERROR` annotations (unsure of their usefulness here).

<details>

```diff
{
    "$message_type": "diagnostic",
    "children": [
        {
            "children": [
            ],
            "code": null,
            "level": "note",
            "message": "config `windows` is only supposed to be controlled by `--target`",
            "rendered": null,
            "spans": [
            ]
        },
        {
            "children": [
            ],
            "code": null,
            "level": "note",
            "message": "see <https://github.com/rust-lang/rust/issues/xxxxx> for more information",
            "rendered": null,
            "spans": [
            ]
        },
        {
            "children": [
            ],
            "code": null,
            "level": "note",
            "message": "`#[deny(unexpected_builtin_cfgs)]` on by default",
            "rendered": null,
            "spans": [
            ]
        }
    ],
    "code": {
        "code": "unexpected_builtin_cfgs",
        "explanation": null
    },
    "level": "error",
    "message": "unexpected `--cfg windows` flag",
    "rendered": "error: unexpected `--cfg windows` flag\n   |\n   = note: config `windows` is only supposed to be controlled by `--target`\n   = note: see <https://github.com/rust-lang/rust/issues/xxxxx> for more information\n   = note: `#[deny(unexpected_builtin_cfgs)]` on by default\n\n",
    "spans": [
+        {
+            "byte_end": 0,
+            "byte_start": 0,
+            "column_end": 1,
+            "column_start": 1,
+            "expansion": null,
+            "file_name": "[..]/rust/tests/ui/cfg/disallowed-cli-cfgs.rs",
+            "is_primary": true,
+            "label": null,
+            "line_end": 1,
+            "line_start": 1,
+            "suggested_replacement": null,
+            "suggestion_applicability": null,
+            "text": [
+            ]
+        }
    ]
}
```

</details>

Conceptually idk if it makes sense to have the diag attached to the crate since at the  point we lint the crate doesn't even exist yet. I also find it slightly better to not use a dummy span in diag since as can be seen here it result in some unexpected behaviours, which is probably fine when it's a mistake or when something went wrong but doesn't seem clean IMO.

Another thing, it's currently not handled by the `annotate_snippet` emitter, which currently completely silently ignores any diag if any span is a dummy span, cf. https://github.com/rust-lang/rust/blob/d371d17496f2ce3a56da76aa083f4ef157572c20/compiler/rustc_errors/src/annotate_snippet_emitter_writer.rs#L139-L142

petrochenkov commented on compiler/rustc_session/src/config/cfg.rs:
```diff
@@ -83,6 +87,76 @@ impl<'a, T: Eq + Hash + Copy + 'a> Extend<&'a T> for ExpectedValues<T> {
+    //  - doc
+    //  - doctest
+    //  - miri
+    //  - rustfmt
```
Technically, this lint is about built-in cfgs and not "well known" cfgs, and all the cfgs in the list below are indeed built-in.
So we don't need to explain why we don't lint them (because they are not set by us).

`test` still needs to be explained though.

Urgau commented:
@rustbot ready

petrochenkov commented:
Ok, let's try landing this.
@bors r+

bors commented:
:pushpin: Commit 971cb135485aaac45ee5841fa280baed1fb1a61e has been approved by `petrochenkov`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=petrochenkov 971cb135485aaac45ee5841fa280baed1fb1a61e -->
<!-- homu: {"type":"Approved","sha":"971cb135485aaac45ee5841fa280baed1fb1a61e","approver":"petrochenkov","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:lock: Merge conflict

This pull request and the master branch diverged in a way that cannot be automatically merged. Please rebase on top of the latest master branch, and let the reviewer approve again.

<details><summary>How do I rebase?</summary>

Assuming `self` is your fork and `upstream` is this repository, you can resolve the conflict following these steps:

1. `git checkout disallow-cfgs` *(switch to your branch)*
2. `git fetch upstream master` *(retrieve the latest master)*
3. `git rebase upstream/master -p` *(rebase on top of it)*
4. Follow the on-screen instruction to resolve conflicts (check `git status` if you got lost).
5. `git push self disallow-cfgs --force-with-lease` *(update this PR)*

You may also read [*Git Rebasing to Resolve Conflicts* by Drew Blessing](http://blessing.io/git/git-rebase/open-source/2015/08/23/git-rebasing-to-resolve-conflicts.html) for a short tutorial.

Please avoid the ["**Resolve conflicts**" button](https://help.github.com/articles/resolving-a-merge-conflict-on-github/) on GitHub. It uses `git merge` instead of `git rebase` which makes the PR commit history more difficult to read.

Sometimes step 4 will complete without asking for resolution. This is usually due to difference between how `Cargo.lock` conflict is handled during merge and rebase. This is normal, and you should still perform step 5 to update this PR.

</details>

<details><summary>Error message</summary>

```text
Auto-merging compiler/rustc_session/src/parse.rs
Auto-merging compiler/rustc_session/src/config.rs
Auto-merging compiler/rustc_lint_defs/src/lib.rs
CONFLICT (content): Merge conflict in compiler/rustc_lint_defs/src/lib.rs
Auto-merging compiler/rustc_lint_defs/src/builtin.rs
Auto-merging compiler/rustc_lint/src/lints.rs
Auto-merging compiler/rustc_lint/src/context/diagnostics.rs
CONFLICT (content): Merge conflict in compiler/rustc_lint/src/context/diagnostics.rs
Auto-merging compiler/rustc_lint/messages.ftl
Automatic merge failed; fix conflicts and then commit the result.

```

</details>

bors commented:
:umbrella: The latest upstream changes (presumably #125741) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

Urgau commented:
Rebased to fix the conflict. CI passes. Let's try again.

@bors r=@petrochenkov

bors commented:
:pushpin: Commit 4fc8318dcc79154504afcf523da78ab08a278ca8 has been approved by `petrochenkov`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=petrochenkov 4fc8318dcc79154504afcf523da78ab08a278ca8 -->
<!-- homu: {"type":"Approved","sha":"4fc8318dcc79154504afcf523da78ab08a278ca8","approver":"petrochenkov","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

tgross35 commented:
Looks like this might have failed the rollup https://github.com/rust-lang/rust/pull/126974

```
2024-06-26T04:46:03.6441365Z failures:
2024-06-26T04:46:03.6441643Z 
2024-06-26T04:46:03.6442639Z ---- build::user_specific_cfgs_are_filtered_out stdout ----
2024-06-26T04:46:03.6444619Z running `/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/cargo rustc -- --cfg debug_assertions --cfg proc_macro`
2024-06-26T04:46:03.6446121Z thread 'build::user_specific_cfgs_are_filtered_out' panicked at tests/testsuite/build.rs:6579:10:
2024-06-26T04:46:03.6446653Z 
2024-06-26T04:46:03.6447485Z test failed running `/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/cargo rustc -- --cfg debug_assertions --cfg proc_macro`
2024-06-26T04:46:03.6448542Z error: process exited with code 101 (expected 0)
2024-06-26T04:46:03.6449005Z --- stdout
2024-06-26T04:46:03.6449163Z 
2024-06-26T04:46:03.6449525Z --- stderr
2024-06-26T04:46:03.6450316Z    Compiling foo v0.5.0 (/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/tmp/cit/t421/foo)
2024-06-26T04:46:03.6451180Z error: unexpected `--cfg debug_assertions` flag
2024-06-26T04:46:03.6451568Z   |
2024-06-26T04:46:03.6452123Z   = note: config `debug_assertions` is only supposed to be controlled by `-C debug-assertions`
2024-06-26T04:46:03.6452945Z   = note: manually setting a built-in cfg does create incoherent behaviors
2024-06-26T04:46:03.6453672Z   = note: `#[deny(unexpected_builtin_cfgs)]` on by default
2024-06-26T04:46:03.6453998Z 
2024-06-26T04:46:03.6454208Z error: unexpected `--cfg proc_macro` flag
2024-06-26T04:46:03.6454573Z   |
2024-06-26T04:46:03.6455136Z   = note: config `proc_macro` is only supposed to be controlled by `--crate-type proc-macro`
2024-06-26T04:46:03.6456187Z   = note: manually setting a built-in cfg does create incoherent behaviors
2024-06-26T04:46:03.6456595Z 
2024-06-26T04:46:03.6456868Z error: could not compile `foo` (bin "foo") due to 2 previous errors
2024-06-26T04:46:03.6457253Z 
2024-06-26T04:46:03.6457555Z note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
2024-06-26T04:46:03.6457976Z 
2024-06-26T04:46:03.6457983Z 
2024-06-26T04:46:03.6458086Z failures:
2024-06-26T04:46:03.6458404Z     build::user_specific_cfgs_are_filtered_out
2024-06-26T04:46:03.6458688Z 
2024-06-26T04:46:03.6459105Z test result: FAILED. 3167 passed; 1 failed; 188 ignored; 0 measured; 0 filtered out; finished in 221.96s
```

matthiaskrgr commented:
@bors r- rollup=iffy

epage commented:
Is it just me or does the lint name `unexpected_builtin_cfgs` make this sound like a subset of `unexpected_cfgs` lint rather than linting that the user is specifying a cfg that shadows? overrides? tweaks? a built-in.

Urgau commented:
Indeed the lint name being so close might be confusing, but I would also argue it's also kind-of a specialized "subset" of `unexpected_cfgs` where we lint on `--cfg` but only for builtin cfgs; which technically could also be done with the `unexpected_cfgs` lint if not for the `RUSTFLAGS` env preventing us to "in general" lint on `--cfg` args.

Of course it's not as simple as this lint on every "possible" builtin cfgs, even "well known" in the sense of the `unexpected_cfgs` lint, but I think it's fine.

petrochenkov commented:
@rfcbot merge

lcnr commented on compiler/rustc_lint_defs/src/lib.rs:
```diff
@@ -747,14 +747,19 @@ pub enum BuiltinLintDiag {
 pub struct BufferedEarlyLint {
     /// The span of code that we are linting on.
-    pub span: MultiSpan,
+    pub span: Option<MultiSpan>,
```
ah, explained further down :sweat_smile:

lcnr commented:
wrt to the lint name, what do you think about `explicit_builtin_cfgs` or `manual_builtin_cfgs`?

jyn514 commented:
> they both seems fine; I have a slight preference for `explicit_builtin_cfgs` bc "manual" doesn't mean much without context

> I though of using something like this `unexpected_builtin_cfgs_in_flags` (or `unexpected_builtin_cfgs_in_cli_flags`) but that feels to verbose

i feel like "unexpected" is not true? the problem is that it's *incoherent*, like i said in the MCP - the cfg doesn't match the behavior of the compiler.

how about `incoherent_cfg_flags`?

Urgau commented:
it's not incoherent to set `unix` on an Linux based target, it's just useless since it's already set, so it feels a bit weird to say "incoherent" in those cases

as written the lint doesn't differentiate between coherent and incoherent behaviour and just disallow setting any built-ins cfg via `--cfg`, mainly for simplicity of implementation, but also bc no-one should be manually setting a built-in cfg

but in that vein I think `explicit_builtin_cfgs_in_flags` goes to point (while letting to door open to allow coherent cfgs to not be linted about if desired)

lcnr commented:
:+1: to `explicit_builtin_cfgs_in_flags`

michaelwoerister commented:
Do all the configs that are not allowed anymore already produce an error? Or is this likely to cause compilation errors for code that currently builds?

Urgau commented:
> Do all the configs that are not allowed anymore already produce an error?

They don't currently produce an error or warning, that's the issue. Someone doing `rustc --target x86_64-unknown-linux-gnu --cfg windows` won't (currently) see any errors or warnings, but this invocation is bogus as it it will active `unix` and `windows` cfg, which is certainly incoherent and not expected.

> Or is this likely to cause compilation errors for code that currently builds?

No code should be affected (we only lint on `--cfg` flags). Only custom build systems, `cargo::rustc-cfg` or `RUSTFLAGS` users might be affect and only **if** they manually set a built-in cfg. I did a GitHub Search when starting this PR and except for the `test` cfg, I didn't find any evidence of users setting a built-in cfg. Given this and the fact that it requires some somewhat advanced `rustc` usage, I'm therefore not expecting the lint to fire likely.

michaelwoerister commented:
> but this invocation is bogus as it it will active unix and windows cfg

Right. I'm trying to find out if any code or build configuration that is working now would start breaking (e.g. because it was not actually using anything affected by the incoherent configs). The proposal of the PR and MCP make sense to me, I'm just wondering how much breakage this will cause. From your comments it sounds like it shouldn't be much. But maybe a crater run would make sense so that we are not caught by surprise if it does affect more projects than expected?

Urgau commented:
> But maybe a crater run would make sense so that we are not caught by surprise if it does affect more projects than expected?

I didn't do an crater run because I don't think it will find anything, as crater only run `cargo check/build/doc` which would only detect `cargo::rustc-cfg` in `build.rs`, but wouldn't detect custom `RUSTFLAGS` or custom build system, which IMO are more likely to set built-in cfg.

Urgau commented:
@estebank @jackh726 @pnkfelix @wesleywiser
There are 2 missing checkboxes for [this FCP](https://github.com/rust-lang/rust/pull/126158#issuecomment-2196371697), could two of you review the FCP.

bors commented:
:umbrella: The latest upstream changes (presumably #125443) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

Urgau commented:
@bors r=@petrochenkov

bors commented:
:pushpin: Commit 20090e6a735441930cc068bb4179b247adcc5c0b has been approved by `petrochenkov`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=petrochenkov 20090e6a735441930cc068bb4179b247adcc5c0b -->
<!-- homu: {"type":"Approved","sha":"20090e6a735441930cc068bb4179b247adcc5c0b","approver":"petrochenkov","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 20090e6a735441930cc068bb4179b247adcc5c0b with merge 968d0ac01a6bbefa3b4de00b28cc7fa1ace2e412...
<!-- homu: {"type":"BuildStarted","head_sha":"20090e6a735441930cc068bb4179b247adcc5c0b","merge_sha":"968d0ac01a6bbefa3b4de00b28cc7fa1ace2e412"} -->

rust-log-analyzer commented:
The job **`x86_64-gnu-aux`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/10240319968/job/28326996100) [(plain)](https://github.com/rust-lang-ci/rust/commit/968d0ac01a6bbefa3b4de00b28cc7fa1ace2e412/checks/28326996100/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

```

</details>

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/10240319968/job/28327989034)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/actions/runs/10240319968/job/28327989034","builder_name":"checks-actions"} -->

petrochenkov commented:
Blocked on https://github.com/rust-lang/cargo/pull/14352.
@rustbot blocked

Urgau commented:
The cargo submodule update was merged in https://github.com/rust-lang/rust/pull/128763, which included https://github.com/rust-lang/cargo/pull/14352, unblocking this pr.

@rustbot labels -S-blocked
@bors r=@petrochenkov

bors commented:
:bulb: This pull request was already approved, no need to approve it again.

- This pull request previously failed. You should add more commits to fix the bug, or use `retry` to trigger a build again.
- There's another pull request that is currently being tested, blocking this pull request: #128768

bors commented:
:pushpin: Commit 20090e6a735441930cc068bb4179b247adcc5c0b has been approved by `petrochenkov`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=petrochenkov 20090e6a735441930cc068bb4179b247adcc5c0b -->
<!-- homu: {"type":"Approved","sha":"20090e6a735441930cc068bb4179b247adcc5c0b","approver":"petrochenkov","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 20090e6a735441930cc068bb4179b247adcc5c0b with merge dd4b6954654fe663c361ce472730a8f3aa48efc2...
<!-- homu: {"type":"BuildStarted","head_sha":"20090e6a735441930cc068bb4179b247adcc5c0b","merge_sha":"dd4b6954654fe663c361ce472730a8f3aa48efc2"} -->

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/10278915062/job/28443461155)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/actions/runs/10278915062/job/28443461155","builder_name":"checks-actions"} -->

rust-log-analyzer commented:
A job failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/10278915062/job/28443378216) [(plain)](https://github.com/rust-lang-ci/rust/commit/dd4b6954654fe663c361ce472730a8f3aa48efc2/checks/28443378216/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

```

</details>

tgross35 commented:
Networking failure

@bors retry

bors commented:
:hourglass: Testing commit 20090e6a735441930cc068bb4179b247adcc5c0b with merge a198c035739972de920f4ad6c6a54f365a251215...
<!-- homu: {"type":"BuildStarted","head_sha":"20090e6a735441930cc068bb4179b247adcc5c0b","merge_sha":"a198c035739972de920f4ad6c6a54f365a251215"} -->

rust-log-analyzer commented:
The job **`i686-gnu-nopt`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/10280863229/job/28449132735) [(plain)](https://github.com/rust-lang-ci/rust/commit/a198c035739972de920f4ad6c6a54f365a251215/checks/28449132735/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
test [ui] tests/ui/cfg/disallowed-cli-cfgs.rs#overflow_checks_ ... ok
test [ui] tests/ui/cfg/diagnostics-not-a-def.rs ... ok
test [ui] tests/ui/cfg/disallowed-cli-cfgs.rs#panic_ ... ok
test [ui] tests/ui/cfg/diagnostics-reexport.rs ... ok
test [ui] tests/ui/cfg/disallowed-cli-cfgs.rs#proc_macro_ ... ok
test [ui] tests/ui/cfg/diagnostics-same-crate.rs ... ok
test [ui] tests/ui/cfg/disallowed-cli-cfgs.rs#relocation_model_ ... ok
test [ui] tests/ui/cfg/disallowed-cli-cfgs.rs#sanitize_ ... ok
test [ui] tests/ui/cfg/disallowed-cli-cfgs.rs#sanitizer_cfi_generalize_pointers_ ... ok
test [ui] tests/ui/cfg/disallowed-cli-cfgs.rs#sanitizer_cfi_normalize_integers_ ... ok
test [ui] tests/ui/cfg/disallowed-cli-cfgs.rs#target_abi_ ... ok
test [ui] tests/ui/cfg/disallowed-cli-cfgs.rs#target_endian_ ... ok
test [ui] tests/ui/cfg/disallowed-cli-cfgs.rs#target_arch_ ... ok
test [ui] tests/ui/cfg/disallowed-cli-cfgs.rs#target_feature_ ... ok
test [ui] tests/ui/cfg/disallowed-cli-cfgs.rs#target_env_ ... ok
test [ui] tests/ui/cfg/disallowed-cli-cfgs.rs#target_family_ ... ok
test [ui] tests/ui/cfg/disallowed-cli-cfgs.rs#target_family_ ... ok
test [ui] tests/ui/cfg/disallowed-cli-cfgs.rs#target_has_atomic_equal_alignment_ ... ok
test [ui] tests/ui/cfg/disallowed-cli-cfgs.rs#target_has_atomic_ ... ok
test [ui] tests/ui/cfg/disallowed-cli-cfgs.rs#target_os_ ... ok
test [ui] tests/ui/cfg/disallowed-cli-cfgs.rs#target_has_atomic_load_store_ ... ok
test [ui] tests/ui/cfg/disallowed-cli-cfgs.rs#target_pointer_width_ ... ok
test [ui] tests/ui/cfg/disallowed-cli-cfgs.rs#ub_checks_ ... ok
test [ui] tests/ui/cfg/disallowed-cli-cfgs.rs#target_vendor_ ... ok
test [ui] tests/ui/cfg/disallowed-cli-cfgs.rs#target_thread_local_ ... ok
test [ui] tests/ui/cfg/disallowed-cli-cfgs.rs#unix_ ... ok
test [ui] tests/ui/cfg/disallowed-cli-cfgs.rs#windows_ ... ok
test [ui] tests/ui/cfg/future-compat-crate-attributes-using-cfg_attr.rs ... ok
test [ui] tests/ui/check-cfg/allow-macro-cfg.rs ... ok
test [ui] tests/ui/check-cfg/allow-at-crate-level.rs ... ok
test [ui] tests/ui/check-cfg/allow-top-level.rs ... ok
---
failures:

---- [assembly] tests/assembly/x86-return-float.rs#unix stdout ----

error in revision `unix`: compilation failed!
status: exit status: 1
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/i686-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/assembly/x86-return-float.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/i686-unknown-linux-gnu/stage2" "--target=i686-unknown-linux-gnu" "--cfg" "unix" "--check-cfg" "cfg(FALSE,unix,windows)" "-Cdebug-assertions=no" "--emit" "asm" "-C" "prefer-dynamic" "-o" "/checkout/obj/build/i686-unknown-linux-gnu/test/assembly/x86-return-float.unix/x86-return-float.s" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/i686-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/i686-unknown-linux-gnu/test/assembly/x86-return-float.unix/auxiliary" "-Ctarget-feature=+sse2" "-Ctarget-cpu=pentium4" "-O" "-C" "force-frame-pointers"
--- stderr -------------------------------
error: unexpected `--cfg unix` flag
  |
  = note: config `unix` is only supposed to be controlled by `--target`
```

</details>

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/10280863229/job/28451518349)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/actions/runs/10280863229/job/28451518349","builder_name":"checks-actions"} -->

Urgau commented:
Another case of a `revisions:` containing `windows`/`unix` as revision name.
Fixed those and re-did a grep on `tests/`.

But just to be sure, and to not waste more CI-merge time. Let's do a big try build across many different targets.
@bors try

bors commented:
:hourglass: Trying commit c0c57b3e2953241124810a59faf77d50e91b3e01 with merge ac7969db6cc18caaabd206a34b8c5034a329765d...
<!-- homu: {"type":"TryBuildStarted","head_sha":"c0c57b3e2953241124810a59faf77d50e91b3e01","merge_sha":"ac7969db6cc18caaabd206a34b8c5034a329765d"} -->

petrochenkov commented:
r=me after CI is green.
@rustbot author

lcnr commented:
@bors r=petrochenkov

bors commented:
:pushpin: Commit c0c57b3e2953241124810a59faf77d50e91b3e01 has been approved by `petrochenkov`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=petrochenkov c0c57b3e2953241124810a59faf77d50e91b3e01 -->
<!-- homu: {"type":"Approved","sha":"c0c57b3e2953241124810a59faf77d50e91b3e01","approver":"petrochenkov","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

Urgau commented:
Wait, try isn't finishing yet, and bors doesn't like r+ and try at the same time.
@bors r-

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/10284379061/job/28468078274)
Build commit: ac7969db6cc18caaabd206a34b8c5034a329765d (`ac7969db6cc18caaabd206a34b8c5034a329765d`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/10284379061/job/28468078274"},"merge_sha":"ac7969db6cc18caaabd206a34b8c5034a329765d"} -->

Urgau commented:
@bors r=petrochenkov

bors commented:
:pushpin: Commit c0c57b3e2953241124810a59faf77d50e91b3e01 has been approved by `petrochenkov`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=petrochenkov c0c57b3e2953241124810a59faf77d50e91b3e01 -->
<!-- homu: {"type":"Approved","sha":"c0c57b3e2953241124810a59faf77d50e91b3e01","approver":"petrochenkov","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit c0c57b3e2953241124810a59faf77d50e91b3e01 with merge ce20e15f01a9dcc4543e8d52b4b281df6c4dbb8e...
<!-- homu: {"type":"BuildStarted","head_sha":"c0c57b3e2953241124810a59faf77d50e91b3e01","merge_sha":"ce20e15f01a9dcc4543e8d52b4b281df6c4dbb8e"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/10288991628/job/28481923609)
Approved by: petrochenkov
Pushing ce20e15f01a9dcc4543e8d52b4b281df6c4dbb8e to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"petrochenkov","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/10288991628/job/28481923609"},"merge_sha":"ce20e15f01a9dcc4543e8d52b4b281df6c4dbb8e"} -->

rust-timer commented:
Finished benchmarking commit (ce20e15f01a9dcc4543e8d52b4b281df6c4dbb8e): [comparison URL](https://perf.rust-lang.org/compare.html?start=8d0066922b14cee3175e8c141e5e909fa6d9a6eb&end=ce20e15f01a9dcc4543e8d52b4b281df6c4dbb8e&stat=instructions:u).

### Overall result: no relevant changes - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=8d0066922b14cee3175e8c141e5e909fa6d9a6eb&end=ce20e15f01a9dcc4543e8d52b4b281df6c4dbb8e&stat=instructions:u)
This benchmark run did not return any relevant results for this metric.

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=8d0066922b14cee3175e8c141e5e909fa6d9a6eb&end=ce20e15f01a9dcc4543e8d52b4b281df6c4dbb8e&stat=max-rss)
<details>
<summary>Results (primary -2.6%, secondary -1.9%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | 3.3%  | [3.3%, 3.3%]   | 1     |
| Improvements ✅ <br /> (primary)   | -2.6% | [-2.6%, -2.6%] | 1     |
| Improvements ✅ <br /> (secondary) | -3.6% | [-3.7%, -3.5%] | 3     |
| All ❌✅ (primary)                 | -2.6% | [-2.6%, -2.6%] | 1     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=8d0066922b14cee3175e8c141e5e909fa6d9a6eb&end=ce20e15f01a9dcc4543e8d52b4b281df6c4dbb8e&stat=cycles:u)
This benchmark run did not return any relevant results for this metric.

### [Binary size](https://perf.rust-lang.org/compare.html?start=8d0066922b14cee3175e8c141e5e909fa6d9a6eb&end=ce20e15f01a9dcc4543e8d52b4b281df6c4dbb8e&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 760.558s -> 760.891s (0.04%)
**Artifact size**: 336.96 MiB -> 336.96 MiB (0.00%)
